src/query/interpreter.cpp

```diff
@@ -1909,6 +1923,138 @@ PreparedQuery PrepareSettingQuery(ParsedQuery parsed_query, const bool in_explic
+    InterpreterContext *interpreter_context, const std::optional<std::string> &username, bool isAdmin) {
+  std::vector<std::vector<TypedValue>> results;
+  results.reserve(interpreter_context->interpreters->size());
+  interpreter_context->interpreters.WithLock([&results, username, isAdmin](const auto &interpreters) {
+    for (const Interpreter *interpreter : interpreters) {
+      std::optional<uint64_t> transaction_id = interpreter->GetTransactionId();
+      if (transaction_id.has_value() && (interpreter->username_ == username || isAdmin)) {
+        auto queries_summaries_typed = std::vector<TypedValue>();
+        auto queries_summaries = interpreter->GetQueries();
+        std::for_each(queries_summaries.begin(), queries_summaries.end(),
+                      [&queries_summaries_typed](const auto &query_summary) {
+                        queries_summaries_typed.emplace_back(query_summary);
+                      });
+        results.push_back({TypedValue(interpreter->username_.value_or("")),
+                           TypedValue(std::to_string(transaction_id.value())), TypedValue(queries_summaries_typed)});
+      }
+    }
+  });
+  return results;
+}
+
+std::vector<std::vector<TypedValue>> TransactionQueueQueryHandler::KillTransactions(
+    InterpreterContext *interpreter_context, const std::vector<std::string> &maybe_kill_transaction_ids,
+    const std::optional<std::string> &username, bool isAdmin) {
+  std::vector<std::vector<TypedValue>> results;
+  for (const std::string &transaction_id : maybe_kill_transaction_ids) {
+    bool killed = false;
+    interpreter_context->interpreters.WithLock([&transaction_id, &killed, username, isAdmin](const auto &interpreters) {
+      for (Interpreter *interpreter : interpreters) {
+        // linearize from this point
+        // the problem can happen e.g: you see that transaction is alive you check transaction_id and username but
+        // interpreter->AbortTransactionByUser() kills the new transaction even if the new transaction hasn't started,
+        // as soon as it starts it will be killed We can try two approaches:
+        // 1. somehow make sure that they are executed consecutively, not de-scheluded from the cpu, so all atomic
+        // I think this would correspond to having lock on the Interpreter
+        // 2. don't create new transaction while we are in the process of killing this one
+
+        // the assumption is that this whole block should be executed atomically
+        std::optional<uint64_t> intr_trans = interpreter->GetTransactionId();
+        if (intr_trans.has_value() && std::to_string(intr_trans.value()) == transaction_id) {  // but this can change
+          if (interpreter->username_ == username ||
+              isAdmin) {  // this cannot change between executions of two transactions in the same interpreter
+            TransactionStatus loaded_status = TransactionStatus::ALIVE;
```

Comment:
What's the semantic of `NO_TRANSACTION`?

Reply:
That interpreter is idle, doesn't execute anything.